### PR TITLE
subsys: mgmt: Fix broken OTA firmware update

### PR DESCRIPTION
**Pull request by MariuszSkamra**
[DNM] ext: mcumgr: Refactor req & rsp buffer freeing
    
    This refactors the buffer free mechaninsm, so that application
    do not have to free the rsp buffer from tx context.
    
    Relates to https://github.com/apache/mynewt-mcumgr/pull/8
    
    Fixes 8636.
    Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

### CODE REVIEW

**Comment by carlescufi** [comment]
CC @JoeHut

File: subsys/mgmt/smp_bt.c
Comment (jhedberg) [inline subsys/mgmt/smp_bt.c]: This is broken. You have to use bt_conn_ref() whenever making persistent references to a bt_conn (preferably when making any reference)

File: subsys/mgmt/smp_bt.c
Comment (jhedberg) [inline subsys/mgmt/smp_bt.c]: No need for a separate local `conn` variable here

File: subsys/mgmt/smp_bt.c
Comment (jhedberg) [inline subsys/mgmt/smp_bt.c]: No need for a local `conn` variable here. Just use bt_smp_conn directly.

File: subsys/mgmt/smp_bt.c
Comment (jhedberg) [reply subsys/mgmt/smp_bt.c]: Also, you're missing the place where you'd release the reference. If you don't get the reference counting right you'll be stuck with a seemingly valid (non-NULL) pointer but you'll never know when it's actually pointing to a valid object. The only way to ensure this is to grab a reference yourself.

**Review by jhedberg — CHANGES_REQUESTED** [review]
I also commented in your own tree, since that's what I got an email notification for first. I only later discovered that you had made a pull request as well.

File: subsys/mgmt/smp_bt.c
Comment (nvlsianpu) [reply subsys/mgmt/smp_bt.c]: I wonder whether it is  good idea to take the reference only for offset == 0? AFIK It depends form client side implementation of the long write operation.

**Comment by MariuszSkamra** [comment]
@jhedberg @nvlsianpu 
I think that conn reference should be per buffer. Managing reference on connected/disconnected is not the right approach in my opinion, because I do not know how it would work in case of multiple connections. 
I've pushed the final version of this patch. I've tested basic image listing, update, listing again, works.
If the user_data is transport specific, the transport has to have an option to free the user_data IMO. 

If you have another vision of this, I will not be mad if you push the patch.

File: subsys/mgmt/smp_bt.c
Comment (jhedberg) [inline subsys/mgmt/smp_bt.c]: Since you know that the full user data is of type smp_bt_user_data, I'd avoid this memcpy and just directly do:

```
if (src_ud->conn) {
        dst_ud->conn = bt_conn_ref(src_ud->conn);
}
```

I think that would be the clean way to do this.

**Comment by nvlsianpu** [comment]
@MariuszSkamra Looks good to me. I've tested this as well and it works perfectly.
Please fix <b>Shippable.</b>

**Review by jhedberg — CHANGES_REQUESTED** [review]
This looks good to me, except for the user data copying part that I've commented on. It'd be important to get a code owner to do the final review however.

**Comment by MariuszSkamra** [comment]
@ccollins476ad Could you please take a look?

**Comment by carlescufi** [comment]
>This looks good to me, except for the user data copying part that I've commented on

@nvlsianpu can you take a look at the part @jhedberg mentions?

**Review by nvlsianpu — APPROVED** [review]
As @jhedberg wrote - it is correct now.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
